### PR TITLE
[Merged by Bors] - chore: move `Polynomial.basisMonomials` to a separate file

### PR DESCRIPTION
**Pull request by acmepjz**
... and generalize it to `Semiring`.

Also add `Polynomial.toFinsuppIsoLinear` which is even true for `Semiring` and is in the very definition of `Polynomial` file.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Polynomial/Basis.lean [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/-
+Copyright (c) 2024 Jz Pan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jz Pan
+-/
+import Mathlib.Algebra.Polynomial.Basic
+import Mathlib.LinearAlgebra.Basis.Defs
```
Comment (alreadydone) [inline Mathlib/Algebra/Polynomial/Basis.lean:7]: Maybe you could import Mathlib.Algebra.Polynomial.Basic intno Mathlib.LinearAlgebra.Basis.Defs instead. Trans imports went up from 649 to 655.

Comment (acmepjz) [reply Mathlib/Algebra/Polynomial/Basis.lean]: Looks good. But let's wait for other's opinions.

Comment (acmepjz) [reply Mathlib/Algebra/Polynomial/Basis.lean]: No that's not possible:
```text
error: ././././Mathlib/RingTheory/Adjoin/Basic.lean:25:18: Declaration Polynomial is not allowed to be imported by this file.
It is defined in Mathlib.Algebra.Polynomial.Basic,
  which is imported by Mathlib.LinearAlgebra.Basis.Defs,
  which is imported by this file.

These invariants are maintained by `assert_not_exists` statements, and exist in order to ensure that "complicated" parts of the library are not accidentally introduced as dependencies of "simple" parts of the library.
```
Let's fallback to separate file design.

Comment (alreadydone) [reply Mathlib/Algebra/Polynomial/Basis.lean:7]: Thanks for testing! `assert_not_exists` is the only thing I was unsure about.

Comment (alreadydone) [reply Mathlib/Algebra/Polynomial/Basis.lean:7]: I think if you hard reset to [671fa8b](https://github.com/leanprover-community/mathlib4/pull/19933/commits/671fa8b094c266b4b8911edff51a4a0dbdbf4b84) and then force push, then the CI won't need to run again.

**Review by alreadydone — APPROVED** [review]
maintainer merge

File: Mathlib/Algebra/Polynomial/Basis.lean (hunk 1 above)
Comment (alreadydone) [reply Mathlib/Algebra/Polynomial/Basis.lean:7]: Actually, it seems CI didn't build mathlib again according to [the log](https://github.com/leanprover-community/mathlib4/actions/runs/12314177521/job/34369737226?pr=19933). So maybe there's no real gain with the hard-reset approach apart from a cleaner history.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**